### ISSUE TITLE
Possible fixes for Makefile

### ISSUE THREAD

**Issue by rpgoldman** [atlas-engineer/nyxt#300]
Here is a diff for the Makefile that contains a number of candidate fixes:

1. Replace use of `load` with the asd file with `asdf:load-asd`.  In modern ASDF, this is the correct way to load system definitions; standard `load` should not be used.  Note that loading next.asd with a relative pathname caused issues with prove-asdf, so I used an absolute pathname.
2. Pre-load next with `ql:quickload`.  When I tried to load normally, ASDF would fail over and over because I was missing various CL libraries. By quickloading next first, I ensured that all of the files were present.
3.  Candidate fix for #298 

```
diff --git a/Makefile b/Makefile
index 04fd369..f9ce504 100644
--- a/Makefile
+++ b/Makefile
@@ -27,7 +27,8 @@ next: $(lisp_files)
 		--eval '(when (string= (uiop:getenv "NEXT_INTERNAL_QUICKLISP") "true") (load "$(QUICKLISP_DIR)/setup.lisp"))' \
 		--eval '(ql:quickload :trivial-features)' \
 		--eval '(ql:quickload :prove-asdf)' \
-		--load next.asd \
+		--eval "(asdf:load-asd \"${PWD}/next.asd\")" \
+		--eval '(ql:quickload "next")' \
 		--eval '(asdf:make :next)' \
 		--eval '(uiop:quit)'
 
@@ -37,7 +38,9 @@ app-bundle: next
 	mkdir -p ./Next.app/Resources
 	mv ./next ./Next.app/Contents/MacOS
 	cp -r ./ports/pyqt-webengine/ ./Next.app/Contents/MacOS
-	mv ./Next.app/Contents/MacOS/next-pyqt-webengine.py ./Next.app/Contents/MacOS/next-pyqt-webengine
+	mkdir ./Next.app/Contents/MacOS/next-pyqt-webengine
+	cp ./ports/pyqt-webengine/next-pyqt-webengine.py ./Next.app/Contents/MacOS/next-pyqt-webengine
+	# mv ./Next.app/Contents/MacOS/next-pyqt-webengine.py ./Next.app/Contents/MacOS/next-pyqt-webengine
 	chmod +x ./Next.app/Contents/MacOS/next-pyqt-webengine
 	cp ./assets/Info.plist ./Next.app/Contents
 	cp ./assets/next.icns ./Next.app/Resource
```

**Comment by jmercouris** [comment]
I will try this out later, thank you :)

**Comment by vindarel** [comment]
This looks good to me. Don't you want to send a PR ?
Thanks !

**Comment by rpgoldman** [comment]
I didn't know how to send a PR (since I haven't forked the repo -- I just installed with homebrew, which was easier), so I am attaching a patch file.

Commit messages corresponding to this patch are:

> commit fba38a8a1c037b8a7dca5318737d4284a0d65aa4 (HEAD -> Makefile-patch)
> Author: Robert P. Goldman <rpgoldman@goldman-tribe.org>
> Date:   Sat Aug 10 15:36:11 2019 -0500

>  Improve interaction with ASDF.
    
>    .asd files should be loaded with ASDF:LOAD-ASD, not CL:LOAD.  Also modify the pathname, because LOAD-ASD wants an absolute pathname.
 >   Pre-load "next" system with QL:QUICKLOAD before ASDF:MAKE.  If libraries are being loaded from quicklisp, they won't be found by ASDF:MAKE.

> commit dc42d376fab83ce3806d4b9eaea8f97e98fdf2e2
> Author: Robert P. Goldman <rpgoldman@goldman-tribe.org>
> Date:   Sat Aug 10 15:34:02 2019 -0500

>  Fix issue with next-pytq-webengine.
    
>  The move command does not seem to work, because the source file is not where expected. Also, a directory didn't exist when expected.

[makefile-fixes.patch.txt](https://github.com/atlas-engineer/next/files/3489275/makefile-fixes.patch.txt)

**Comment by seanfarley** [comment]
The above patch is just plain wrong. It makes a directory and then `chmod +x`? That gives this error:

```
<ERROR> [15:07:55] next base.lisp (initialize-port remote-interface fun24) -
  Couldn't execute "./Next.app/Contents/MacOS/next-pyqt-webengine": Permission denied
```

I can't tell what the original desire was but it seems `cp -r ./ports/pyqt-webengine/ ./Next.app/Contents/MacOS` should be `cp -r ./ports/pyqt-webengine/* ./Next.app/Contents/MacOS`.

Plus, `./assests/Info.plist` misspells `next` as `Next` for which are two different names on a case-sensitive filesystem.

**Comment by rpgoldman** [comment]
@seanfarley I'm actually not sure what you are talking about here.

1. The `chmod +x` was already in the `Makefile`; I didn't do anything to that.
2. I didn't have any interaction with `Info.plist` at all.

So in none of these ways was I "just plain wrong."  The original Makefile may have had issues that I didn't fix, but complaining about my patch is really off-base.

I assume you didn't intend to, but your tone read to me as quite rude, especially since your comments were off-base: you seem to have misinterpreted the context lines of the diff as my own code.  I can find projects to contribute to where people are more pleasant.

**Comment by seanfarley** [comment]
Apologies for coming off rude; that wasn't my intention. Yes, I understand you didn't write the `chmod` line but if you had run it then you would have seen that putting the executable inside the directory of the same is not correct.

As for leaving based on my interactions: I'm not even part of this project. Just a rando dude trying to get this to work on the mac.

**Comment by vindarel** [comment]
Yep, thanks for the explanation. Us three of the team (me, jmercouris, ambrevar) are glad you send suggestions of improvements. And I'll leave the review to my colleagues for the coming days : ]

**Comment by Ambrevar** [comment]
@jmercouris: Do you think we can merge this?

**Comment by Ambrevar** [comment]
@jmercouris Friendly ping! :)

**Comment by Ambrevar** [comment]
Is this still valid?

**Comment by jmercouris** [comment]
I am not sure!

**Comment by seanfarley** [comment]
I can try to build this again; last time I tried was a year or so ago.

**Comment by seanfarley** [comment]
Ok, I've test this out and this issue is no longer applicable. Still doesn't build on macos because webkitgtk isn't a thing and qt6 doesn't have webengine anymore, so I'm at a lost of how to even get this to work on mac at all.

**Comment by Ambrevar** [comment]
Oh, I didn't realize that QtWebEngine was not in Qt6.  Any idea if it
will be re-added?

**Comment by seanfarley** [comment]
> Oh, I didn't realize that QtWebEngine was not in Qt6. Any idea if it will be re-added?

No idea, I don't follow Qt development but it seems it's just flatout gone now: https://doc.qt.io/qt-6/whatsnew60.html

**Comment by seanfarley** [comment]
I spent today trying to get the MacPorts version working as another approach. (I used to be a MacPorts maintainer so I thought I'd try that route.) There were a few hurdles to get the dependencies to build. It's a bit unfortunate that gtk is used; it works but there is a definite lag.

These are just a brain dump of notes, mostly to myself, in case I want to make some PRs later; so don't worry too much about them if they're too nitpicky.

Hurdles:
* whack-a-mole with getting gtk installed without x11
* [neon has a build bug](https://github.com/notroj/neon/issues/45) (workaround: `sudo /opt/local/bin/port install neon configure.cflags=-Wno-error=all`)
* had to edit webkit-gtk port to use python39 for its build dep since pulling in python2 busts py-gobject stuff (I'm not sure about commiting that to MacPorts, though)
* a few libraries, like libepoxy, still pull in X deps which is slightly annoying [1]
* had to edit libproxy to add python39 variant since it's not using the python portgroup
* at this point, I could move on to building nyxt
* edited the nyxt portfile to depend on webkit2-gtk since, historically, the `-devel` ports aren't updated as frequently
* edited the nyxt portfile to set build.jobs to 1, I couldn't get it to compile in parallel (seems that the makefile tried to build the Nyxt.app bundle too early and failed on a file not yet compiled)

All in all, not the worst project I've made work with macports (that honor goes to MUMPS and MPI). As I mentioned before, the drawing and keyboard input is a bit laggy (I can only assume due to all the socket+gtk interaction on a mac). There are also these weird WebKitWebProcess processes that show up in the dock as well (screenshot below):

<img width="1440" alt="Screen Shot 2021-03-25 at 10 57 12 PM" src="https://user-images.githubusercontent.com/20887/112576992-b2e45100-8db0-11eb-9d44-d34e74e60950.png">

Anyways, that's probably enough for today. Feel free to close this issue since it's definitely not applicable anymore.

[1] I ended up with these X libraries:
```
xorg-libice                    @1.0.10_0 
xorg-libpthread-stubs          @0.4_0 
xorg-libsm                     @1.2.3_0 
xorg-libX11                    @1.7.0_0 
xorg-libXau                    @1.0.9_0 
xorg-libxcb                    @1.14_0+python39 
xorg-libXdamage                @1.1.5_0 
xorg-libXdmcp                  @1.1.3_0 
xorg-libXext                   @1.3.4_0 
xorg-libXfixes                 @5.0.3_1 
xorg-libXi                     @1.7.10_0 
xorg-libXmu                    @1.1.3_0 
xorg-libXt                     @1.2.1_0 
xorg-libXtst                   @1.2.3_1 
xorg-libXxf86vm                @1.1.4_1 
xorg-xcb-proto                 @1.14.1_0+python39 
xorg-xorgproto                 @2021.3_0
```

**Comment by Ambrevar** [comment]
@jmercouris What are you thoughts on this?  Can we still get the MacPorts in better shape after all?

**Comment by Ambrevar** [comment]
@seanfarley Thank you so much for this incredible investigation!

**Comment by jmercouris** [comment]
There are also some patches I've added to get rid of the multitude of "WebkitWebProcess" icons in the dock. I believe that we could get these merged upstream as well. It would be interesting to see master running on macOS now  too :-D.

**Comment by seanfarley** [comment]
I just saw that building webkit-gtk with quartz [doesn't have opengl](https://github.com/macports/macports-ports/blob/36a4990f197af54e31881f5e5a59b880c51116dd/www/webkit2-gtk/Portfile#L137) which might explain some of the laginess.

What would the path be to use native macos webkit? I assume it would be to build a `cl-macos-webkit` library to directly bind to webkit?

**Comment by jmercouris** [comment]
The makefile has mostly been superseded by the ASD. Please feel free to reopen if you notice any more problems!
